interp/iostream.go

```diff
@@ -0,0 +1,203 @@
+package interp
+
+import (
+	"bufio"
+	"io"
+	"os/exec"
+	"syscall"
+)
+
+const (
+	notClosedExitCode = -127
+)
+
+func cmdExitCode(cmdError error, ioerrs... error) (exitCode int, err error) {
```

Comment:
I think I should really add more comments. This function was part of a refactor to remove the common logic from closing the `exec.Cmd` for both input and output. It is confusing but that is because the _stdlib_ error handling is complicated.

I can merge this function into `exitErrorCode` but I thought it would be less overwhelming to separate them into two phases so that `exitErrorCode` only receives an `*exec.ExitError`. The problem is that `cmdError` might be an `io` error that occurred when closing the `exec.Cmd`. Additionally, there might have been other `io` errors that occurred when flushing the pipe _to_ the `exec.Cmd` before `Wait()`-ing.

On top of that, an `*exec.ExitError` will occur if the child exit `!= 0` but this is already reported by the `ExitCode()`, and so it is discarded. This logic is separated into the `exitErrorCode()` func. That allows any `io` error prior to `Wait()` to bubble up and replace it.

There are two goals here:
1. The command must be `Wait()`-ed on even if an `io` error occurs before the `Wait()` call. Otherwise,  resources are not freed in the `goawk` process for the child.
2. Avoid hiding any errors that occur while fulfilling Goal 1. Some hiding may occur but the "worst" error is always returned.

I should make it more explicit but it is possible for the exit code to indicate success but also return an error code for `io` errors!

### Truth table

Here is a truth table of all possible error conditions. A `Wait()` error that reports a non-zero exit code or signal exit of a child is equal to `nil`.

| Struct | Prior I/O Error | `Wait()` error | Exit Code Result | Error Result |
| ------ | -------------- | ------------- | ---------- | ----- |
| inCmdStream | nil | nil | >= 0 | nil |
| " | nil | IOWaitError | -1 | IOWaitError |
| " | CloseError | nil | >= 0 | CloseError |
| " | CloseError | IOWaitError | -1 | IOWaitError |
| outCmdStream | nil | nil | >= 0 | nil |
| " | nil | IOWaitError | -1 | IOWaitError |
| " | FlushError | nil | >= 0 | FlushError |
| " | FlushError | IOWaitError | -1 | IOWaitError |
| " | CloseError | nil | >= 0 | CloseError |
| " | CloseError | IOWaitError | -1 | IOWaitError |
| " | FlushError + CloseError | nil | >= 0 | FlushError |
| " | FlushError + CloseError | IOWaitError | -1 | IOWaitError |

Reply:
Thanks for the effort on your comment and the truth table. However, I don't think "more comments" or a big truth table will really help here -- and it may just make it more text to wade through and so harder to understand. I think the reason it's hard to follow now is because you're trying to abstract too much: handle three different cases, with multiple errors as input, and then the `goto` and bare returns don't help clarity.

So here's a concrete suggestion that I think will work and (to me) is very clear and simple: How about we move only the `exec.ExitError` check into `exitErrorCode`, and have it take a regular `error`, like so:

```go
func exitErrorCode(cmdError error) (exitCode int, err error) {
	if ee, ok = cmdError.(*exec.ExitError); !ok {
		return -1, err
	}
	// ... rest of exitErrorCode
}
```

And then make the three cases that use it concrete, boring, and clear by just spelling out the errors they check. Like so:

```go
// system()
err = cmd.Wait()
if err != nil {
	// ... check ctx
	code, err := exitErrorCode(err)
	ret = float64(code)
	if err != nil {
		p.printErrorf("%v\n", err)
	}
}

func (s *outCmdStream) Close() (err error) {
	flushErr := s.Writer.Flush()
	closeErr := s.Closer.Close()
	cmdErr := s.cmd.Wait()
	code, err := exitErrorCode(cmdErr)
	s.exitCode = code
	if err != nil {
		return err
	}
	if flushErr != nil {
		return flushErr
	}
	if closeErr != nil { // could just "return closeErr" here, but I think this is slightly clearer
		return closeErr
	}
	return nil
}

func (s *inCmdStream) Close() error {
	closeErr := s.ReadCloser.Close()
	cmdErr := s.cmd.Wait()
	code, err := exitErrorCode(cmdErr)
	s.exitCode = code
	if err != nil {
		return err
	}
	if closeErr != nil {
		return closeErr
	}
	return nil
}
```

Reply:
Oh ok I see what you are saying now. That sounds good to simplify.

Reply:
I've uplifted the error checking into a separate function:

```go
// firstError returns the first non-nill error or nil if all errors are nil.
func firstError(errs ...error) error {
    for _, err := range errs {
        if err != nil {
            return err
        }
    }
    return nil
}
```

Now it is no longer necessary to pass in the previous errors to the `cmdExitCode` function and it is much simpler (i.e. no `goto` statements). I've also changed the `cmdExitCode` function to `waitExitCode` and perform the `.Wait()` inside of it, because a `.Wait()` always preceded calling `cmdExitCode`.

Now I can use `firstError` in three places of `iostream.go` and avoid using it for the `system()` builtin function. I think it makes the error priority much more obvious, on a single line, compared to reading through 10 lines of boilerplate:

```go
	s.exitCode, waitErr = waitExitCode(cmd)
	return firstError(cmdErr, flushErr, closeErr)
	// VS
	if cmdErr != nil {
		return cmdErr
	}
	if flushErr != nil {
		return flushErr
	}
	if closeErr != nil {
		return closeErr
	}
	return nil
```